packages/source-iottwinmaker/src/knowledgeGraph-module/KGDataModule.ts

```diff
@@ -0,0 +1,58 @@
+import {
+  ExecuteQueryCommand,
+  ExecuteQueryCommandOutput,
+  ExecuteQueryCommandInput,
+  IoTTwinMakerClient,
+  ColumnDescription,
+  Row,
+} from '@aws-sdk/client-iottwinmaker';
+import { KGDataModuleInput, executeQueryParams } from './types';
+import { TwinMakerKGQueryDataModule } from '../types';
+
+const MAX_RESULTS_PER_PAGE = 25;
+
+export class KGDataModule implements TwinMakerKGQueryDataModule {
+  private twinMakerClient: IoTTwinMakerClient;
+  private workspaceId: string;
+
+  constructor(input: KGDataModuleInput) {
+    this.twinMakerClient = input.twinMakerClient;
+    this.workspaceId = input.workspaceId;
+  }
+
+  executeQuery = async ({
+    queryStatement,
+    resultsPerPage,
+    maxPagesCount,
+  }: executeQueryParams): Promise<ExecuteQueryCommandOutput> => {
+    const rows: Row[] = [];
+    let columnDescriptions: ColumnDescription[] = [];
+    let $metadata = {};
+    const commandInput: ExecuteQueryCommandInput = {
+      workspaceId: this.workspaceId,
+      queryStatement: queryStatement,
+      maxResults: resultsPerPage,
+    };
+    do {
+      const response: ExecuteQueryCommandOutput = await this.twinMakerClient.send(
+        new ExecuteQueryCommand(commandInput)
+      );
+      if (response.rows) {
+        rows.push(...response.rows);
+      }
+      if (response.columnDescriptions && columnDescriptions.length === 0) {
+        columnDescriptions = response.columnDescriptions;
+      }
+      if (response.$metadata && Object.keys($metadata).length === 0) $metadata = response.$metadata;
+      commandInput.nextToken = response.nextToken;
+    } while (
+      maxPagesCount
+        ? commandInput.nextToken &&
+          (resultsPerPage && resultsPerPage > 0
+            ? rows.length / resultsPerPage < maxPagesCount
+            : rows.length / MAX_RESULTS_PER_PAGE < maxPagesCount)
```

Comment:
I think you should have a counter for the number of pages received instead of using the rows to calculate. 

this calculation relies on the assumptions that for each previous pages, the backend will always return the max number of rows, and also the default max result is always 25 and won't change in the future. These assumptions make this calculation easy to break.

Reply:
Hmm, These assumptions are based on API contracts I feel it is a stretch to account for if API contract is changed. The backend should always return the max number of rows except for the last page if there is no more data and this case is guarded by the fact there will be no next token. However I like the idea of page counter here is how the code will look like let me know 

do {
      const response: ExecuteQueryCommandOutput = await this.twinMakerClient.send(
        new ExecuteQueryCommand(commandInput)
      );
      pageCounter = pageCounter + 1;
      if (response.rows) {
        rows.push(...response.rows);
      }
      if (response.columnDescriptions && columnDescriptions.length === 0) {
        columnDescriptions = response.columnDescriptions;
      }
      if (response.$metadata && Object.keys($metadata).length === 0) $metadata = response.$metadata;
      commandInput.nextToken = response.nextToken;
    } while (maxPagesCount ? commandInput.nextToken && pageCounter < maxPagesCount : commandInput.nextToken);
    return { rows, columnDescriptions, $metadata };
  };

Reply:
yes